server/routerlicious/packages/services/src/throttler.ts

```diff
@@ -0,0 +1,113 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IThrottler,
+    IThrottlerResponse,
+    IThrottleManager,
+    ThrottlerRequestType,
+    IRequestMetrics,
+} from "@fluidframework/server-services-core";
+
+/**
+ * Implements the Token Bucket algorithm for throttling requests.
+ */
+export class Throttler implements IThrottler {
+    constructor(
+        private readonly throttleManager: IThrottleManager,
+        private readonly requestRate: number,
+        private readonly minCooldownIntervalInMs: number,
+    ) {
+    }
+
+    public async updateRequestCount(
+        id: string,
+        requestType: ThrottlerRequestType,
```

Comment:
If possible, I would also like to keep this class generic and keep the requestType in Helper class.

Also should we switch the class names? ThrottlerHelper should be the inner class and Throttler uses that class?

Reply:
I think that makes sense. To that point, (after swapping classNames) should the names be more accurate to their implementation specifics? For example, this is a "Lenient" Throttler using a "TokenBucket" ThrottlerHelper, so should the classes be LenientThrottler and TokenBucketThrottlerHelper? This would leave room, for example StrictThrottler or LeakyBucketThrottlerHelper, if someone had different requirements but wanted to use the same IThrottler and IThrottlerHelper structure

Reply:
ThrottlerHelper implementation wouldn't change with rating algorithms, would it?
The helper class is just called Throttler and maintaining in-mem dictionary?

Reply:
No, the 2 wouldn't affect the other and could be swapped around as needed. 

The scenario I'm thinking of is where someone cares more about accuracy than latency, so they implement `StrictThrottler` which `await`s `ThrottlerHelper.updateRequestCount()` or `ThrottlerHelper.getRequestCount()`, rather than allowing them to run in the background and cache when finished like `LenientThrottler`. Both would use `TokenBucketThrottler`.

Alternative scenario would be someone wants the pros/cons of, say, fixed window rate-limiting instead of token bucket. Here, they would implement `FixedWindowThrottlerHelper`, but could use `LenientThrottler` or `StrictThrottler` depending on the requirements.